### DISCUSSION TITLE
Data directory defined in user-defined external config json file

### DISCUSSION THREAD

**Discussion by JohnsonDylan** [pyinstaller/pyinstaller#7377] [Help]
I apologize in advance if I've overlooked where this has been covered previously. 

A script I am writing reads a user config file (userconfig.json) that is located in the same directory as the script file. That config file is used to define the absolute path to a directory where their data is stored (containing one or more ~1GB files). This data directory can be anywhere in their filesystem. This file is being located in the script using the following:
"""
filename = "userconfig.json" 
contents = open(filename).read()
config = eval(contents)
file_directories = config['file_directories']
"""
In this way, the userconfig.json file is not packaged with pyinstaller. Upon creation of the exe, that userconfig file is dropped into the directory containing the exe. The exe is generated by pyinstaller without error and the userconfig file is dropped into the same directory as the exe. 

Upon running the exe, the script loads the userconfig file which points to the users defined data directory. Running the standalone python script works correctly and essentially follows the workflow: 0) iterate over user-defined data directory 1) convert datafile into different format and save to same data directory 2) load newly created data file as pandas dataframe 3) run 1 or more queries on the pandas dataframe and append the results of each query to a result dataframe 4) repeat

I have so far tested this with a data directory containing only two datafiles. This will eventually scale up to 50+ files. Running the pyinstaller exe results in one of three different outcomes, even when nothing has changed between runs)

1) The program works fine (though I suspect it will not work fine once I scale up)
2) The program ends abruptly without any indication of failure. This occurs during either step 1 or 2 of the above workflow and can occur with either the first or second file in the data directory. No downstream code that saves the resultant output is executed.
3) A windows error window pops up stating "The instruction at ### reference memory at ###. The memory could not be read."

I see a lot of documentation about adding data files within an imported package/module to the installer itself (and have successfully done this to my knowledge as I bypassed a previous error), but I have not come across documentation on the appropriate way to handle user-defined data input and suspect that I am maybe loading the data incorrectly. 

I had originally been running the executable with my working directory being the same directory as the executable. I have also modified the code to where my working directory is where the data is stored and that produced the same result.

Thanks for any assistance or suggestions

**Comment by rokm** [comment]
> ```python
> filename = "userconfig.json"
> contents = open(filename).read()
> config = eval(contents)
> file_directories = config['file_directories']
> ``` 

For starters, this `userconfig.json` should be referred to with an absolute path (e.g., anchored to executable's location), otherwise you are implicitly assuming that it is located in the current working directory. And surely a JSON file should be parsed using `json.loads()` instead of being `eval`'d?

> Running the pyinstaller exe results in one of three different outcomes, even when nothing has changed between runs)
> 
>   1. The program works fine (though I suspect it will not work fine once I scale up)
>   2. The program ends abruptly without any indication of failure. This occurs during either step 1 or 2 of the above workflow and can occur with either the first or second file in the data directory. No downstream code that saves the resultant output is executed.
>   3. A windows error window pops up stating "The instruction at ### reference memory at ###. The memory could not be read."

This sounds a lot like an undefined behavior caused by un-initialized pointer deep in the bowels of your processing code (i.e., the C/C++ extension code of either `pandas` or `numpy`), well beyond ours or your reach. What are the typical values of both ###? Is this perhaps data-induced (e.g., malformed data files) - does it happen with all data files that you have, or with specific ones?

It's hard to say anything else without seeing the code and exemplar data, and being able to reproduce the crash, though. Can you provide a minimal example? Trying to come up with one might actually also help you isolate the root of the problem...

**Reply by JohnsonDylan** [reply]
Thank you for the reply. 
I've corrected the points you made regarding the path and loading of the json file.
In trying to come up with a minimal example, I realized that, although the data files being referenced to in the userconfig.json file are approximately 1 GB each, I had not realized the resultant pandas dataframes from opening those files are approximately **2.5** GB each (from df.memory_usage(index=True).sum())

I admittedly don't know much about issues with memory limits, but I suspect this might be the problem. 
(My PC is 64bit, with 96 GB of RAM however)

I will revise my original description to exclude the outcome where the program worked fine. I'm now consistently getting the error message and I'm unable to replicate the instance where it worked. 

The windows popup error windows states: "The instruction at 0x00007FFD097FD4CB referenced memory at 0x0000000181199440. The memory could not be read."

It's quite difficult to create a minimal example to replicate this. I have found that if I bypass the call to the module that handles the process of creating the ~2.5 GB pandas dataframe and feed in a dummy smaller dataframe, the issue is not present.

**Reply by JohnsonDylan** [reply]
Okay, some progress. There seem to be two different issues. One is in the pandas dataframe memory usage and the other is in accessing cached data files.

As to the former, I'll have to verify there are no repercussions, but I deleted the data with values of zero and was able to reduce the memory usage of the dataframes by half. Now the program is working under one of two scenarios (at least with my current dataset). I'm not sure if this is exactly a long term solution, as there may be datasets that produce dataframes that are ~2.5 GB even with the zero values removed. Anyway, I will investigate this further and see if there's anywhere else I can reduce the dataframe sizes.

Regarding the latter issue of accessing cached data files, a package I'm using looks in the user-defined data directory outside of the pyinstaller package and iterates over the files. If the 'cache' parameter of the module is True, for each file, the package checks to see if there is a cached feather (pyarrow) version of the file. That module calls this line: `if os.path.exists(filename)`. If I run python version of my script, the file is found, but in the pyinstaller exe, the program crashes at this line and I get the windows memory error pop-up. If I `print(os.path)` from my python script, I get `<module 'ntpath' from 'C:\\Users\\<username>\\AppData\\Local\\Programs\\Python\\Python310\\lib\\ntpath.py'>`

My understanding is that this path is invalid in the pyinstaller exe

**Reply by JohnsonDylan** [reply]
For anyone coming across this, I removed UPX compression and now I'm longer experiencing the memory issues described above.
